### ISSUE TITLE
v0.2.3: streamed (layer-major) prefill produces NaN in the MoE output on RTX 3090 / WSL2, then SIGSEGV; cache-batched path is correct

### ISSUE THREAD

**Issue by sittingmongoose** [Apolog1ze-Dev/QwFNfer#8]
## Environment
- QwFNfer v0.2.3 release bundle `qwfnfer-linux-x86_64-cuda.zip` (SHA256 `683a1e0d1ecee688a3b697c83eb15df441509bed5f14296a95b0144e86e18b21`), source tag `887dacd`
- Model: `unsloth/Qwen3.8-Flash-Next-GGUF` rev `38bb39ee97821de2c9009abb7e93950eec396e66`, `UD-Q4_K_XL` (4 shards)
- GPU: RTX 3090 24 GB (sm_86), NVIDIA driver 610.88 (Windows host), also driving the desktop
- CPU: Ryzen 9 5950X (Zen 3, no AVX-512); 32 GB RAM host, WSL2 VM memory limit 24 GB
- OS: Ubuntu 24.04.4 under WSL2 (kernel 6.6.87.2-microsoft-standard-WSL2), model on the distro's ext4 disk.
  io_uring and O_DIRECT both work there (console probe: 6.8 GB/s random 2 MiB reads)

## Reproduction (minimal)
```
qwfn-server UD-Q4_K_XL-00001-of-00004.gguf --port 8081 --ctx 32768 --kv f16 --batch 2048 --ram 9 --ram-frac 0.85 \
  --threads 8 --reserve 1792 --think xhigh --think-budget 6000 --no-prefill-overlap
```
(no --spec-block, no --mtp, no --state-host, no --mmproj; same result with the console's tuned 128K plan)

1. A 33-token chat request answers correctly.
2. A ~1.2K-token request (synthetic log lines plus one planted sentence, `reasoning_effort: off`, `max_tokens: 64`)
   kills the server with SIGSEGV before the first generated token. Generator (Python):
   `random.Random(1042)`; 33 lines `[rec {i:06d}] service=<choice> host=node-<1..400> latency_ms=<3..2400> ticket OPS-<1000..99999>`;
   one line inserted at index 11; question appended.

## Observations
- Fatal trace is host code, identical in every run (tuned 128K plan, 32K minimal plan, ~1K and ~20K prompts):
  `qwfn-server(+0x8c857) <- +0x4abde <- +0x6d735 <- +0x6de8b <- +0xef47d <- +0xf3225 <- +0xf41c6 <- +0xcc95f`
  (signal-handler frames `+0x401ca`/`+0x40280` above it). No CUDA frames in the fatal stack.
- `QWFN_TRACE_PREFILL=1` (with `QWFN_LEGACY_PREFILL=1`): MoE input finite at every layer up to the first bad one,
  then the MoE **output** goes non-finite and it propagates to layer 47; SIGSEGV follows after the last layer:
  - run 1: `layer 5 moe_IN nonfinite=0` -> `layer 5 moe_out sum=nan nonfinite=20480` (8 token rows of 2560)
  - run 2: `layer 3 moe_out sum=nan nonfinite=762880` (298 rows) — the first bad layer is not fixed
- With the default (non-legacy) streamed path on a 30K-token document the prefill completed (350 tok/s) and decoding
  produced `!!!!!!…` (token id 0 repeated) instead of crashing — consistent with NaN logits.
- `QWFN_CHECK_STRIDE=1`: 144/144 expert ranges OK (no MISMATCH).
- `QWFN_VERIFY_UPLOAD=1`: every layer `mismatching slices=0 of 1536`, worst pointer alignment 4096 — device copies equal
  the host staging, even with the verification's synchronisation, and the NaN still appears.
- `QWFN_VRAM_AUDIT=1`: 6335 MB free after init; no pagefile writes observed during the failing runs.
- Non-streamed paths are correct on the identical request (2 runs each, plus known-answer follow-ups):
  - `--batch 64 --prefill-decode-max 64` (cache-batched chunks): correct, ~12.8 tok/s prompt processing
  - same + `QWFN_NO_CACHE_BATCH=1` (token by token): correct, ~11.7 tok/s
- The STALL watchdog fires during long prefills at `generated token 0` even when the request later completes; it is not
  the crash (the ~1.2K case segfaults with no STALL at all).

## Summary of tested execution paths
| Path | How selected | Result on the ~1.2K request |
|---|---|---|
| Streamed, default (`eval_prefill_big`) | default flags, batch 2048 | SIGSEGV (and `!!!!` output on a 30K prompt that completed) |
| Streamed, legacy per-ubatch | `QWFN_LEGACY_PREFILL=1` | non-finite MoE output (layer 5; layer 3 in a second run), then SIGSEGV |
| Cache-batched | `--batch 64 --prefill-decode-max 64` | correct, 2/2 runs plus known-answer follow-ups |
| Token by token | same + `QWFN_NO_CACHE_BATCH=1` | correct |

- Both tested streamed-prefill implementations failed; the cache-batched and token-by-token paths passed every case tested
  (~1K x3, 4K and 14K needles with follow-ups, one image, one structured tool round trip).
- Non-finite MoE outputs preceded the host-side crash in every traced run.
- The `STALL` watchdog probe stacks (`ggml_cuda_op_top_k` ... `cuStreamSynchronize`) were not the fatal stacks; the fatal
  stack is the host-only one above, and the ~1.2K case crashed with no STALL at all.
- Synchronized host-staging-to-GPU byte comparisons passed on every layer.

## What is not established
- Staged host bytes were not independently compared with the source file bytes.
- Contexts above ~14K tokens and end-to-end use through a client harness were not verified on the working paths.
- Host/GPU memory was not sampled during the working-path progression tests.

Whether the bad values originate in the host staging bytes (streamed O_DIRECT/io_uring reads through a WSL2 virtual disk)
or in the batched MoE kernels on sm_86: VERIFY_UPLOAD compares device against staging, not staging against the file.
A staging-vs-file check (or a CPU-side recompute of the first bad layer's MoE) would separate the two.

## Independent check of the model file
The same four `UD-Q4_K_XL` shards (SHA256 matching the Hugging Face listing) run correctly on native Windows llama.cpp b11039
(CUDA 13.4, same GPU): ~1K/4K/14K needle prompts with follow-ups, an image and a structured tool round trip all answered
correctly. So the file itself is not implicated — though that run changes both the runtime and the OS.

**Comment by Apolog1ze-Dev** [comment]
I'll take a look at it asap . I'm just swamped so it might take a bit of time

**Comment by sittingmongoose** [comment]
Thank you!  It’s preventing me from using it sadly and with my low(32gb) ram, it’s the only config that would work for me haha
